File: src/main/java/ch/bubendorf/gsak2gpx/SqlTemplateMethod.java
package ch.bubendorf.gsak2gpx;

import freemarker.template.TemplateMethodModelEx;
import freemarker.template.TemplateModel;
import freemarker.template.TemplateModelException;
import freemarker.template.utility.DeepUnwrap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.List;

@SuppressWarnings("WeakerAccess")
public class SqlTemplateMethod implements TemplateMethodModelEx {

    private final Logger LOGGER = LoggerFactory.getLogger(SqlTemplateMethod.class.getSimpleName());

    private final Connection connection;

    public SqlTemplateMethod(Connection connection) {
        this.connection = connection;
    }

    @Override
    public Object exec(List arguments) throws TemplateModelException {
        if (arguments.size() < 1 ) {
            throw new TemplateModelException("Missing argument: sql");
        }
        String sql = arguments.get(0).toString();
        if (arguments.get(0) instanceof TemplateModel) {
            final Object unwrap = DeepUnwrap.unwrap((TemplateModel) arguments.get(0));
            sql = unwrap.toString();
        }
        LOGGER.debug(sql);
        String category = arguments.size() < 2 ? "" : arguments.get(1).toString();

        try {
            final Statement statement = connection.createStatement();
            final ResultSet rs = statement.executeQuery(sql);

            return new ResultSetCollectionModel(rs, category);
        } catch (SQLException e) {
            throw new TemplateModelException(e);
        }
    }
}

File: src/main/kotlin/ch/bubendorf/gsak2gpx/SqlTemplateMethod.kt
package ch.bubendorf.gsak2gpx

import freemarker.template.TemplateMethodModelEx
import freemarker.template.TemplateModel
import freemarker.template.TemplateModelException
import freemarker.template.utility.DeepUnwrap
import org.slf4j.LoggerFactory
import java.sql.Connection
import java.sql.SQLException

class SqlTemplateMethod(private val connection: Connection) : TemplateMethodModelEx {

    private val LOGGER = LoggerFactory.getLogger(SqlTemplateMethod::class.java.simpleName)

    @Throws(TemplateModelException::class)
    override fun exec(arguments: List<*>): Any {
        if (arguments.size < 1) {
            throw TemplateModelException("Missing argument: sql")
        }
        var sql = arguments[0].toString()
        if (arguments[0] is TemplateModel) {
            val unwrap = DeepUnwrap.unwrap(arguments[0] as TemplateModel)
            sql = unwrap.toString()
        }
        LOGGER.debug(sql)
        val category = if (arguments.size < 2) "" else arguments[1].toString()

        try {
            val statement = connection.createStatement()
            val rs = statement.executeQuery(sql)

            return ResultSetCollectionModel(rs, category)
        } catch (e: SQLException) {
            throw TemplateModelException(e)
        }

    }
}
